src/ports/postgres/modules/utilities/pivot.py_in

```diff
@@ -58,66 +61,255 @@ def pivot(schema_madlib, source_table, out_table,
+        param_types['__madlib_def_pval__'] = list
+        agg_dict = extract_keyvalue_params(aggregate_func,param_types)
+    except KeyError, e:
+        with MinWarning("warning"):
+            plpy.warning("Pivot: Not all columns from '{aggregate_func}'"
+            " present in '{pivot_values}'".format(**locals()))
+            raise
+
+    # If the dictionary is empty, parse it as a list
+    agg_set = split_quoted_delimited_str(aggregate_func) if len(agg_dict) < 1 \
+         else []
+    # If the list is empty set 'avg' as default
+    agg_dict['__madlib_def_agg__'] = ['avg'] if len(agg_set) < 1 else agg_set
 
     # Find the distinct values of pivot_cols
-    distinct_values = plpy.execute(
-        "SELECT array_agg(DISTINCT {pcol} ORDER BY {pcol}) AS value "
-        "FROM {source_table}".
-        format(pcol=pcol[0], source_table=source_table))
-
-    distinct_values = [strip_end_quotes(item)
-                        for item in distinct_values[0]['value']]
-    # The aggregate collects pivot_values values for a given pivot_cols value
-    case_str = ("{agg}("
-                "CASE WHEN \"{{pcol}}\" = '{{value}}' THEN {pval} ELSE NULL END"
-                ")".
-                format(agg=aggregate_func,
-                       pval=pval_no_quotes))
-    sql_list.append(
-        ", " +
-        # Assign the name of the new column
-        ', '.join("{case_str} as \"{{pcol}}_{{value}}\"".
-                  format(case_str=case_str).
-                  format(pcol=pcol_no_quotes, value=str(value))
-                  for value in distinct_values if value is not None))
+    # Note that the distinct values are appended in order of users list
+    # This ordering is important when we access pivot_comb entries
+    distinct_str = ["SELECT "]
+    distinct_str.append(
+        ', '.join("array_agg(DISTINCT {pcol}) AS {pcol}_values".
+            format(pcol=pcol) for pcol in pcols))
+    distinct_str.append(" FROM " + source_table)
+    distinct_values = plpy.execute(''.join(distinct_str))
+
+    # Collect the distinc values for every pivot column
+    pcol_distinct_values = {}
+    pcol_max_length = 0
+    for pcol in pcols:
+        # Read the distinct values for this pcol
+        pcol_tmp = [item for item in distinct_values[0][pcol+"_values"]]
+        # Remove null values if keep null is not true
+        if not keep_null:
+            pcol_tmp = [x for x in pcol_tmp if x is not None]
+        pcol_distinct_values[pcol]=pcol_tmp
+        # Max length of the string that pcol values can create +
+        # length of pcol + 1 (for _ character)
+        pcol_max_length+=max([len(str(item)) for item in pcol_tmp])+len(pcol)+1
+
+    # For non postgres systems add null to the pcol_distinct_values if necessary
+    # This is required because DISTINCT keyword is implemented differently
+    is_pg = m4_ifdef(<!__POSTGRESQL__!>, True, False)
+    if keep_null and not is_pg:
+        null_str_begin = ["SELECT "]
```

Comment:
Again, this would be cleaner with the whole query as a single string. 
Also you could simplify it by using `EXISTS (SELECT 1 FROM {tbl} WHERE {pcol} is NULL) as {p_col}_isnull, ...`